### DISCUSSION TITLE
implementing AXIS CGI commands using Python Requests with username+password authentication

### DISCUSSION THREAD

**Discussion by thilleary** [AxisCommunications/discussions#37] [VAPIX] [answered]
Can anyone provide a Python example of a successful http GET request to implement a CGI command?
For example, the following code tries to enable dynamic text overlay (sending a "#D"):

import urllib3
http = urllib3.PoolManager()
resp = http.request('GET', 'http://192.168.1.179:9079/axis-cgi/param.cgi?action=update&Image.I0.Text.TextEnabled=yes&Image.I0.Text.String=%23D')
print(resp.data)
# get the status of the response
print(resp.status)

The code connects with the camera server but because there is no username & password authentication, a 401 error is returned:

b'<HTML><HEAD><TITLE>401 Unauthorized</TITLE></HEAD>\n<BODY><H1>401 Unauthorized</H1>\nYour client does not have permission to get URL /axis-cgi/param.cgi from this server.\n</BODY></HTML>\n'
401

**Comment by mattias-kindborg-at-work** [comment]
You should use [Digest access authentication](https://en.wikipedia.org/wiki/Digest_access_authentication) when communicating with the device. Do you want to give it a go, or would you like me to attempt to provide you with an example?

**Reply by thilleary** [reply]
Thought about using digest, but I saw no reason that basic would not work - it would be a great starting point, whichever, just so see something work. If you can whip up an example I love to see it.
Thanks

**Reply by mattias-kindborg-at-work** [reply]
Basic access authentication over HTTP is not safe, thus devices with newer firmware are configured to only use digest access authentication. You can configure the camera to accept basic access authentication, but I would advice against it.

I'll see if I can create a small sample for you in the upcoming days.

**Comment by thilleary** [comment]
I really appreciate the advice. Here’s the whole picture:

 

Initially, we were able to send a cgi commands to the axis camera (firmware 5.5 or greater) using an Arduino type processor. The processor established an ethernet socket type connection, sent the cgi command, authenticated, and then closed the socket.   

 

We could successfully send virtual input change of state commands to the axis camera: 

"GET /axis-cgi/admin/virtualinput/activate.cgi?schemaversion=1&port=[port#]"

 

Or repeatedly send dynamic text commands to embed in the camera image to reflect that status of a process being monitored visually and with performance metrics:

"GET /axis-cgi/dynamicoverlay.cgi?action=settext&text=[text showing operating metrics]"

Previously the camera had been set up to enable dynamic text operation (#D). 

 

This worked, but it is non-deterministic and portions of the code were empirically derived (they work but for unknown reasons). The objective now is to optimize and move this process to python application running on a Linux platform, and do it the right way.

 

Here is the Arduino code (excerpted) that works adequately (for dynamic text overly):
[see attached pdf]
[Arduino Text Overlay Code.pdf](https://github.com/AxisCommunications/discussions/files/9264467/Arduino.Text.Overlay.Code.pdf)


It seems that we should be able to do this faster and more efficiently using the camera’s VAPIX3 APIs. But authenticating to the camera (username and password) is an elusive puzzle, and I remain unconvinced it can be done using the APIs.

 

So, alternatively, establishing (or better yet persisting) an ethernet connection to the camera and using the process that appeared to work with the Arduino platform is the current objective.

 

 

 

From: Mattias Kindborg ***@***.***>
Reply-To: AxisCommunications/discussions ***@***.***>
Date: Thursday, August 4, 2022 at 2:22 AM
To: AxisCommunications/discussions ***@***.***>
Cc: Tom Hilleary ***@***.***>, Author ***@***.***>
Subject: Re: [AxisCommunications/discussions] implementing AXIS CGI commands using Python Requests with username+password authentication (Discussion #37)

 

Basic access authentication over HTTP is not safe, thus devices with newer firmware are configured to only use digest access authentication. You can configure the camera to accept basic access authentication, but I would advice against it.

I'll see if I can create a small sample for you in the upcoming days.

**Comment by mattias-kindborg-at-work** [comment, marked as answer]
To start off, I'm not using `urllib3` but instead `requests`. I'm hoping that `urllib3` wasn't a hard requirement.

Create the following files in a directory.

**requirements.txt**

```text
requests
```

**list-camera-parameters.py**

```python
import requests
import sys

if len(sys.argv) != 4:
    print("Usage: list-camera-parameters.py <ip> <username> <password>")
    sys.exit(1)

ip = sys.argv[1]
username = sys.argv[2]
password = sys.argv[3]

auth = requests.auth.HTTPDigestAuth(username, password)
r = requests.get("http://{}/axis-cgi/admin/param.cgi?action=list&group".format(ip), auth=auth)

print(r.status_code)
print(r.text)
```

Then create and activate a new virtual environment:

```sh
python3 -m venv env
source env/bin/activate
```

Proceed with installing the required packages:

```sh
python3 -m pip install -r requirements.txt
```

Then finally run the example with the following command:

```sh
python3 list-camera-parameters.py <ip> <username> <password>
```

where `<ip>` is the IP address, `<username>` is the username and `<password>` is the password of the Axis device.

**Comment by thilleary** [comment]
Mattias,

Perfect – just the push I needed to carry forward.

Thanks for so generously lending your obvious expertise.

BTW, where are you located in the world – and are you directly affiliated with AXIS?

Regards,

/T Hilleary

 

 

From: Mattias Kindborg ***@***.***>
Reply-To: AxisCommunications/discussions ***@***.***>
Date: Friday, August 5, 2022 at 4:35 AM
To: AxisCommunications/discussions ***@***.***>
Cc: Tom Hilleary ***@***.***>, Author ***@***.***>
Subject: Re: [AxisCommunications/discussions] implementing AXIS CGI commands using Python Requests with username+password authentication (Discussion #37)

 

To start off, I'm not using urllib3 but instead requests. I'm hoping that urllib3 wasn't a hard requirement.

Create the following files in a directory.

requirements.txt
requests
list-camera-parameters.py

import requests
import sys
 
if len(sys.argv) != 4:
    print("Usage: list-camera-parameters.py <ip> <username> <password>")
    sys.exit(1)
 
ip = sys.argv[1]
username = sys.argv[2]
password = sys.argv[3]
 
auth = requests.auth.HTTPDigestAuth(username, password)
r = requests.get("http://{}/axis-cgi/admin/param.cgi?action=list&group".format(ip), auth=auth)
 
print(r.status_code)
print(r.text)
Then create and activate a new virtual environment:

python3 -m venv env
source env/bin/activate
Proceed with installing the required packages:

python3 -m pip install -r requirements.txt
Then finally run the example with the following command:

python3 list-camera-parameters.py <ip> <username> <password>
where <ip> is the IP address, <username> is the username and <password> is the password of the Axis device.

**Reply by thilleary** [reply]
One quick followup.

We'll be implementing a number of cgi functions, some with varying payloads (e.g. text overlay).
The following text overlay works great:
r = requests.get("http://{}/axis-cgi/dynamicoverlay.cgi?action=settext&text=This%20Works!".format(ip), auth=auth)

But if we wanted to pass the "settext&text=" content as a variable, what would be the most efficient way to build it into the get request?

**Reply by mattias-kindborg-at-work** [reply]
Yes, I'm fortunate enough to be working at Axis Communications with the awesome developers they've got here. Corporate headquarter is located in the south of Sweden.

Regarding your question about formatting strings, it depends on what you optimize for, but it's usually a waste of time. Start with optimizing for readability because the time spend formatting a string is minuscule compared to the time it takes to send a request and receive its response. 

For information regarding formatting strings, see [here](https://www.w3schools.com/python/python_strings_format.asp). For URL encoding your value, see [here](https://docs.python.org/3/library/urllib.parse.html#url-quoting).

**Comment by nikolaev4812** [comment]
Hello everyone!
I'm using the Axis F44 main modul and AXIS F1004 Pinhole Sensor Unit.
The AXIS F1004 Pinhole Sensor Unit is not an IP camera.
If I uderstood correctly this request is used for IP cameras 
`requests.get("http://{}/axis-cgi/dynamicoverlay.cgi?action=settext&text=This%20Works!".format(ip), auth=auth)`

How can I programmatically (Python) impose a dynamically changing text on camera #1 or #2. The AXIS F44 module has 4 video channels

**Comment by nikolaev4812** [comment]
<img width="563" alt="RTP Settings" src="https://github.com/AxisCommunications/discussions/assets/129624467/6e24b24c-e8ad-4ad9-8d81-b8e098b4ce15">

Here's my RTP Settings in AXIS modul F44.
As we can see each channel of the camera has IP adress but those adresses are on a different subnet with AXIS F44(192.168.0.90)
I'm trying to get video stream using this code
<img width="674" alt="Python_source" src="https://github.com/AxisCommunications/discussions/assets/129624467/8a3c1c44-a254-4117-8951-129f6c5fcf57">
